File: sdk/wechat/mics-client-sdk/src/micsClient.ts
import { buildWsUrl } from "./wsUriBuilder";
import { MicsClientOptions, type MicsClientOptions as MicsClientOptionsType } from "./options";
import type {
  ConnectParams,
  MicsClientCallbacks,
  MicsClientState,
  MicsSession,
  SendGroupChatParams,
  SendSingleChatParams
} from "./types";
import {
  AckStatus,
  ClientFrameCodec,
  type MessageAck,
  type MessageDelivery,
  type MessageRequest,
  MessageType,
  type ServerError,
  ServerFrameCodec
} from "./proto/mics_message";
import type { MpSocket } from "./ws/mpSocket";
import { createWxMpSocket } from "./ws/mpSocket";

type SocketFactory = (url: string) => MpSocket;

function nowMs(): number {
  return Date.now();
}

function sleep(ms: number): Promise<void> {
  if (ms <= 0) return Promise.resolve();
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class MicsClient {
  private readonly options: MicsClientOptionsType;
  private readonly socketFactory: SocketFactory;

  private state: MicsClientState = "disconnected";
  private callbacks: MicsClientCallbacks = {};

  private sock: MpSocket | null = null;
  private session: MicsSession | null = null;
  private connectParams: ConnectParams | null = null;

  private connectAckResolve: ((s: MicsSession) => void) | null = null;
  private connectAckReject: ((e: unknown) => void) | null = null;

  private heartbeatTimer: number | null = null;

  private reconnectTask: Promise<void> | null = null;
  private connectedWaiters: Array<() => void> = [];

  private pendingAcks = new Map<string, { resolve: (ack: MessageAck) => void }>();
  private nextMsgId = 0;

  constructor(
    options: MicsClientOptionsType = MicsClientOptions.default(),
    deps?: { socketFactory?: SocketFactory } & MicsClientCallbacks
  ) {
    this.options = options;
    this.socketFactory = deps?.socketFactory ?? ((url) => createWxMpSocket(url));
    this.callbacks = deps ?? {};
  }

  set onStateChanged(cb: ((s: MicsClientState) => void) | undefined) {
    this.callbacks.onStateChanged = cb;
  }
  set onConnected(cb: ((s: MicsSession) => void) | undefined) {
    this.callbacks.onConnected = cb;
  }
  set onDeliveryReceived(cb: ((d: MessageDelivery) => void) | undefined) {
    this.callbacks.onDeliveryReceived = cb;
  }
  set onAckReceived(cb: ((a: MessageAck) => void) | undefined) {
    this.callbacks.onAckReceived = cb;
  }
  set onServerErrorReceived(cb: ((e: ServerError) => void) | undefined) {
    this.callbacks.onServerErrorReceived = cb;
  }

  getState(): MicsClientState {
    return this.state;
  }

  async connect(params: ConnectParams): Promise<MicsSession> {
    if (this.state !== "disconnected") {
      throw new Error("client is not disconnected");
    }

    this.connectParams = params;
    this.setState("connecting");

    const wsUrl = buildWsUrl(params.url, params.tenantId, params.token, params.deviceId);
    const sock = this.socketFactory(wsUrl);
    this.sock = sock;

    const sessionPromise = new Promise<MicsSession>((resolve, reject) => {
      this.connectAckResolve = resolve;
      this.connectAckReject = reject;
    });

    sock.onMessage((bytes) => void this.handleMessage(bytes));
    sock.onClose(() => void this.handleClose());
    sock.onError(() => {
      // rely on close for state changes
    });

    // wait for connect ack
    const timeout = this.options.connectTimeoutMs;
    const session = await Promise.race([
      sessionPromise,
      (async () => {
        await sleep(timeout);
        throw new Error("connect timeout");
      })()
    ]);

    this.session = session;
    this.setState("connected");
    this.startHeartbeat();
    this.callbacks.onConnected?.(session);
    this.signalConnected();
    return session;
  }

  async disconnect(): Promise<void> {
    this.setState("disposing");
    this.stopHeartbeat();
    try {
      this.sock?.close(1000, "dispose");
    } catch {
      // ignore
    }
    this.sock = null;
    this.session = null;
    this.setState("disconnected");
  }

  async sendSingleChat(params: SendSingleChatParams): Promise<MessageAck> {
    const msgId = params.msgId && params.msgId.length > 0 ? params.msgId : this.nextId();
    const { session } = await this.waitConnectedPair();

    const body = await this.prepareOutboundBody(params.msgBody);
    const msg: MessageRequest = {
      tenantId: session.tenantId,
      userId: session.userId,
      deviceId: session.deviceId,
      msgId,
      msgType: MessageType.SINGLE_CHAT,
      toUserId: params.toUserId,
      groupId: "",
      msgBody: body,
      timestampMs: nowMs()
    };

    return this.sendWithRetry(msg);
  }

  async sendGroupChat(params: SendGroupChatParams): Promise<MessageAck> {
    const msgId = params.msgId && params.msgId.length > 0 ? params.msgId : this.nextId();
    const { session } = await this.waitConnectedPair();

    const body = await this.prepareOutboundBody(params.msgBody);
    const msg: MessageRequest = {
      tenantId: session.tenantId,
      userId: session.userId,
      deviceId: session.deviceId,
      msgId,
      msgType: MessageType.GROUP_CHAT,
      toUserId: "",
      groupId: params.groupId,
      msgBody: body,
      timestampMs: nowMs()
    };

    return this.sendWithRetry(msg);
  }

  private async sendWithRetry(msg: MessageRequest): Promise<MessageAck> {
    let ackSettled = false;
    let ackValue: MessageAck | null = null;

    const ackPromise = new Promise<MessageAck>((resolve) =>
      this.pendingAcks.set(msg.msgId, {
        resolve: (ack) => {
          if (ackSettled) return;
          ackSettled = true;
          ackValue = ack;
          resolve(ack);
        }
      })
    );

    let sentAttempts = 0;
    const maxSendAttempts = Math.max(1, this.options.maxSendAttempts);

    try {
      while (sentAttempts < maxSendAttempts) {
        if (ackSettled && ackValue) return ackValue;

        const { sock } = await this.waitConnectedPair();
        msg.timestampMs = nowMs();
        const frameBytes = ClientFrameCodec.encode({ payload: "message", message: msg }).finish();

        try {
          if (!this.sock) {
            throw new Error("ws not ready");
          }
          await sock.send(frameBytes);
          sentAttempts++;
        } catch {
          this.tryStartReconnect();
          continue;
        }

        const ackWait = await Promise.race([
          ackPromise.then((ack) => ({ kind: "ack" as const, ack })),
          sleep(this.options.ackTimeoutMs).then(() => ({ kind: "timeout" as const }))
        ]);

        if (ackWait.kind === "ack") return ackWait.ack;
      }

      return { msgId: msg.msgId, status: AckStatus.FAILED, timestampMs: nowMs(), reason: "ack timeout" };
    } finally {
      this.pendingAcks.delete(msg.msgId);
    }
  }

  private async handleMessage(bytes: Uint8Array): Promise<void> {
    let frame: ReturnType<typeof ServerFrameCodec.decode>;
    try {
      frame = ServerFrameCodec.decode(bytes);
    } catch {
      return;
    }

    switch (frame.payload) {
      case "connectAck": {
        const ack = frame.connectAck;
        if (ack.code !== 1000) {
          this.connectAckReject?.(new Error(`connect rejected: ${ack.code}`));
          return;
        }
        const session: MicsSession = {
          tenantId: ack.tenantId,
          userId: ack.userId,
          deviceId: ack.deviceId,
          nodeId: ack.nodeId,
          traceId: ack.traceId
        };
        this.connectAckResolve?.(session);
        return;
      }
      case "ack": {
        const ack: MessageAck = frame.ack;
        this.pendingAcks.get(ack.msgId)?.resolve(ack);
        this.callbacks.onAckReceived?.(ack);
        return;
      }
      case "delivery": {
        let delivery: MessageDelivery = frame.delivery;
        const crypto = this.options.messageCrypto;
        if (crypto && delivery.message && delivery.message.msgBody?.byteLength) {
          try {
            const dec = await crypto.decrypt(delivery.message.msgBody);
            delivery = { message: { ...delivery.message, msgBody: dec } };
          } catch {
            // best-effort: surface ciphertext if decrypt fails
          }
        }
        this.callbacks.onDeliveryReceived?.(delivery);
        return;
      }
      case "error": {
        this.callbacks.onServerErrorReceived?.(frame.error as ServerError);
        return;
      }
      case "heartbeatPong":
        return;
      default:
        return;
    }
  }

  private handleClose(): void {
    this.stopHeartbeat();
    this.sock = null;
    this.session = null;
    if (this.state === "disposing") return;
    this.tryStartReconnect();
  }

  private setState(state: MicsClientState): void {
    if (this.state === state) return;
    this.state = state;
    this.callbacks.onStateChanged?.(state);
  }

  private startHeartbeat(): void {
    this.stopHeartbeat();
    const interval = this.options.heartbeatIntervalMs;
    if (!interval || interval <= 0) return;

    this.heartbeatTimer = setInterval(() => {
      if (!this.sock) return;
      const frame = ClientFrameCodec.encode({ payload: "heartbeatPing", heartbeatPing: { timestampMs: nowMs() } }).finish();
      void this.sock.send(frame);
    }, interval) as any;
  }

  private stopHeartbeat(): void {
    if (this.heartbeatTimer !== null) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
  }

  private nextId(): string {
    this.nextMsgId++;
    return String(this.nextMsgId);
  }

  private async prepareOutboundBody(plaintext: Uint8Array): Promise<Uint8Array> {
    if (!plaintext || plaintext.byteLength === 0) return new Uint8Array();
    const crypto = this.options.messageCrypto;
    if (!crypto) return plaintext;
    return crypto.encrypt(plaintext);
  }

  private async waitConnectedPair(): Promise<{ sock: MpSocket; session: MicsSession }> {
    while (true) {
      if (this.state === "connected" && this.sock && this.session) {
        return { sock: this.sock, session: this.session };
      }

      if (this.state === "disconnected" || this.state === "disposing") {
        throw new Error("client is not connected");
      }

      if (!this.options.autoReconnect) {
        throw new Error("client is not connected");
      }

      await new Promise<void>((resolve) => this.connectedWaiters.push(resolve));
    }
  }

  private signalConnected(): void {
    const waiters = this.connectedWaiters;
    this.connectedWaiters = [];
    for (const w of waiters) w();
  }

  private tryStartReconnect(): void {
    if (!this.options.autoReconnect) {
      this.setState("disconnected");
      return;
    }
    if (this.reconnectTask) return;
    if (!this.connectParams) {
      this.setState("disconnected");
      return;
    }

    this.setState("reconnecting");
    const p = this.connectParams;

    this.reconnectTask = (async () => {
      let delay = Math.max(0, this.options.reconnectMinDelayMs);
      const max = Math.max(delay, this.options.reconnectMaxDelayMs);

      while (true) {
        try {
          this.connectAckResolve = null;
          this.connectAckReject = null;
          this.session = null;

          const wsUrl = buildWsUrl(p.url, p.tenantId, p.token, p.deviceId);
          const sock = this.socketFactory(wsUrl);
          this.sock = sock;

          const sessionPromise = new Promise<MicsSession>((resolve, reject) => {
            this.connectAckResolve = resolve;
            this.connectAckReject = reject;
          });

          sock.onMessage((bytes) => void this.handleMessage(bytes));
          sock.onClose(() => void this.handleClose());
          sock.onError(() => {});

          const session = await Promise.race([
            sessionPromise,
            (async () => {
              await sleep(this.options.connectTimeoutMs);
              throw new Error("connect timeout");
            })()
          ]);

          this.session = session;
          this.setState("connected");
          this.startHeartbeat();
          this.callbacks.onConnected?.(session);
          this.signalConnected();
          return;
        } catch {
          if (delay > 0) await sleep(delay);
          delay = Math.min(max, delay === 0 ? 50 : delay * 2);
        }
      }
    })().finally(() => {
      this.reconnectTask = null;
    });
  }
}

File: sdk/flutter/mics_client_sdk/lib/src/mics_client.dart
import 'dart:async';
import 'dart:math';
import 'dart:typed_data';

import 'package:fixnum/fixnum.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

import 'crypto/message_crypto.dart';
import 'options.dart';
import 'proto/proto.dart';
import 'types.dart';
import 'ws_uri_builder.dart';

class MicsClient {
  final MicsClientOptions options;

  MicsClient(this.options);

  final _stateController = StreamController<MicsClientState>.broadcast();
  final _connectedController = StreamController<MicsSession>.broadcast();
  final _deliveryController = StreamController<MessageDelivery>.broadcast();
  final _ackController = StreamController<MessageAck>.broadcast();
  final _serverErrorController = StreamController<ServerError>.broadcast();

  Stream<MicsClientState> get stateStream => _stateController.stream;
  Stream<MicsSession> get connectedStream => _connectedController.stream;
  Stream<MessageDelivery> get deliveryStream => _deliveryController.stream;
  Stream<MessageAck> get ackStream => _ackController.stream;
  Stream<ServerError> get serverErrorStream => _serverErrorController.stream;

  MicsClientState _state = MicsClientState.disconnected;
  MicsClientState get state => _state;

  WebSocketChannel? _ws;
  StreamSubscription? _wsSub;

  MicsSession? _session;
  MicsConnectParams? _connectParams;

  Timer? _heartbeatTimer;

  Completer<MicsSession>? _connectAckCompleter;
  final List<Completer<void>> _connectedWaiters = [];

  final Map<String, Completer<MessageAck>> _pendingAcks = {};
  int _nextMsgId = 0;

  Future<void>? _reconnectFuture;

  Future<MicsSession> connect(MicsConnectParams params) async {
    if (_state != MicsClientState.disconnected) {
      throw StateError('client is not disconnected');
    }

    _connectParams = params;
    _setState(MicsClientState.connecting);
    final session = await _connectInternal(params);
    return session;
  }

  Future<void> disconnect() async {
    _setState(MicsClientState.disposing);
    _stopHeartbeat();
    await _closeWs();
    _session = null;
    _setState(MicsClientState.disconnected);
  }

  Future<MessageAck> sendSingleChat({
    required String toUserId,
    required Uint8List msgBody,
    String? msgId,
  }) async {
    final id = (msgId != null && msgId.isNotEmpty) ? msgId : _nextId();
    final pair = await _waitConnectedPair();

    final body = await _prepareOutboundBody(msgBody);
    final msg = MessageRequest(
      tenantId: pair.session.tenantId,
      userId: pair.session.userId,
      deviceId: pair.session.deviceId,
      msgId: id,
      msgType: MessageType.SINGLE_CHAT,
      toUserId: toUserId,
      groupId: '',
      msgBody: body,
      timestampMs: Int64(_nowMs()),
    );

    return _sendWithRetry(msg);
  }

  Future<MessageAck> sendGroupChat({
    required String groupId,
    required Uint8List msgBody,
    String? msgId,
  }) async {
    final id = (msgId != null && msgId.isNotEmpty) ? msgId : _nextId();
    final pair = await _waitConnectedPair();

    final body = await _prepareOutboundBody(msgBody);
    final msg = MessageRequest(
      tenantId: pair.session.tenantId,
      userId: pair.session.userId,
      deviceId: pair.session.deviceId,
      msgId: id,
      msgType: MessageType.GROUP_CHAT,
      toUserId: '',
      groupId: groupId,
      msgBody: body,
      timestampMs: Int64(_nowMs()),
    );

    return _sendWithRetry(msg);
  }

  Future<MessageAck> _sendWithRetry(MessageRequest msg) async {
    final ackCompleter = Completer<MessageAck>();
    _pendingAcks[msg.msgId] = ackCompleter;

    final maxAttempts = max(1, options.maxSendAttempts);
    int sentAttempts = 0;

    try {
      while (sentAttempts < maxAttempts) {
        if (ackCompleter.isCompleted) {
          return ackCompleter.future;
        }

        final pair = await _waitConnectedPair();
        msg.timestampMs = Int64(_nowMs());

        try {
          final frame = ClientFrame(message: msg);
          pair.ws.sink.add(Uint8List.fromList(frame.writeToBuffer()));
          sentAttempts++;
        } catch (_) {
          _tryStartReconnect();
          continue;
        }

        final result = await Future.any([
          ackCompleter.future.then((ack) => ack),
          Future<MessageAck?>.delayed(options.ackTimeout, () => null),
        ]);

        if (result != null) {
          return result;
        }
      }

      return MessageAck(
        msgId: msg.msgId,
        status: AckStatus.FAILED,
        timestampMs: Int64(_nowMs()),
        reason: 'ack timeout',
      );
    } finally {
      _pendingAcks.remove(msg.msgId);
    }
  }

  Future<_ConnectedPair> _waitConnectedPair() async {
    while (true) {
      final ws = _ws;
      final session = _session;
      if (_state == MicsClientState.connected && ws != null && session != null) {
        return _ConnectedPair(ws: ws, session: session);
      }

      if (_state == MicsClientState.disconnected || _state == MicsClientState.disposing) {
        throw StateError('client is not connected');
      }

      if (!options.autoReconnect) {
        throw StateError('client is not connected');
      }

      final waiter = Completer<void>();
      _connectedWaiters.add(waiter);
      await waiter.future;
    }
  }

  void _signalConnected() {
    final waiters = List<Completer<void>>.from(_connectedWaiters);
    _connectedWaiters.clear();
    for (final w in waiters) {
      if (!w.isCompleted) w.complete();
    }
  }

  Future<MicsSession> _connectInternal(MicsConnectParams params) async {
    await _closeWs();
    _session = null;

    final connectAck = Completer<MicsSession>();
    _connectAckCompleter = connectAck;

    final uri = buildWsUri(params.url, params.tenantId, params.token, params.deviceId);

    late final WebSocketChannel channel;
    try {
      channel = WebSocketChannel.connect(uri);
    } catch (e) {
      _connectAckCompleter = null;
      rethrow;
    }

    _ws = channel;
    _wsSub = channel.stream.listen(
      (data) => _handleWsMessage(data),
      onError: (_) {
        // rely on onDone for state changes
      },
      onDone: _handleWsDone,
      cancelOnError: false,
    );

    final session = await Future.any([
      connectAck.future,
      Future<MicsSession>.delayed(options.connectTimeout, () => throw TimeoutException('connect timeout')),
    ]);

    _session = session;
    _setState(MicsClientState.connected);
    _startHeartbeat();
    _connectedController.add(session);
    _signalConnected();
    return session;
  }

  Future<void> _handleWsMessage(dynamic data) async {
    final bytes = _toBytes(data);
    if (bytes.isEmpty) return;

    ServerFrame frame;
    try {
      frame = ServerFrame.fromBuffer(bytes);
    } catch (_) {
      return;
    }

    switch (frame.whichPayload()) {
      case ServerFrame_Payload.connectAck:
        final ack = frame.connectAck;
        if (ack.code != 1000) {
          _connectAckCompleter?.completeError(StateError('connect rejected: ${ack.code}'));
          return;
        }
        _connectAckCompleter?.complete(MicsSession.fromConnectAck(ack));
        return;

      case ServerFrame_Payload.ack:
        final ack = frame.ack;
        final c = _pendingAcks[ack.msgId];
        if (c != null && !c.isCompleted) {
          c.complete(ack);
        }
        _ackController.add(ack);
        return;

      case ServerFrame_Payload.delivery:
        var delivery = frame.delivery;
        final crypto = options.messageCrypto;
        if (crypto != null && delivery.hasMessage() && delivery.message.msgBody.isNotEmpty) {
          try {
            final dec = await crypto.decrypt(Uint8List.fromList(delivery.message.msgBody));
            final msg = delivery.message.deepCopy()..msgBody = dec;
            delivery = MessageDelivery(message: msg);
          } catch (_) {
            // best-effort: surface ciphertext if decrypt fails
          }
        }
        _deliveryController.add(delivery);
        return;

      case ServerFrame_Payload.error:
        _serverErrorController.add(frame.error);
        return;

      case ServerFrame_Payload.heartbeatPong:
      case ServerFrame_Payload.notSet:
        return;
    }
  }

  void _handleWsDone() {
    _stopHeartbeat();
    _ws = null;
    _wsSub?.cancel();
    _wsSub = null;
    _session = null;

    if (_state == MicsClientState.disposing) return;
    _tryStartReconnect();
  }

  void _tryStartReconnect() {
    if (!options.autoReconnect) {
      _setState(MicsClientState.disconnected);
      return;
    }
    if (_reconnectFuture != null) return;
    final params = _connectParams;
    if (params == null) {
      _setState(MicsClientState.disconnected);
      return;
    }

    _setState(MicsClientState.reconnecting);

    _reconnectFuture = () async {
      var delay = options.reconnectMinDelay;
      final maxDelay = options.reconnectMaxDelay;

      while (true) {
        try {
          _connectAckCompleter = null;
          await _connectInternal(params);
          return;
        } catch (_) {
          final ms = delay.inMilliseconds;
          if (ms > 0) {
            await Future.delayed(_withJitter(delay));
          }
          final nextMs = ms == 0 ? 50 : ms * 2;
          delay = Duration(milliseconds: min(maxDelay.inMilliseconds, nextMs));
        }
      }
    }().whenComplete(() {
      _reconnectFuture = null;
    });
  }

  Duration _withJitter(Duration d) {
    if (d.inMilliseconds <= 0) return d;
    final r = Random().nextInt(max(1, d.inMilliseconds ~/ 4));
    return Duration(milliseconds: d.inMilliseconds + r);
  }

  void _startHeartbeat() {
    _stopHeartbeat();
    if (options.heartbeatInterval.inMilliseconds <= 0) return;

    _heartbeatTimer = Timer.periodic(options.heartbeatInterval, (_) {
      final ws = _ws;
      if (ws == null) return;
      try {
        final ping = HeartbeatPing(timestampMs: Int64(_nowMs()));
        final frame = ClientFrame(heartbeatPing: ping);
        ws.sink.add(Uint8List.fromList(frame.writeToBuffer()));
      } catch (_) {
        // ignore
      }
    });
  }

  void _stopHeartbeat() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = null;
  }

  Future<void> _closeWs() async {
    final ws = _ws;
    _ws = null;
    _session = null;
    _connectAckCompleter = null;

    await _wsSub?.cancel();
    _wsSub = null;

    if (ws != null) {
      try {
        await ws.sink.close(1000, 'dispose');
      } catch (_) {
        // ignore
      }
    }
  }

  void _setState(MicsClientState s) {
    if (_state == s) return;
    _state = s;
    _stateController.add(s);
  }

  String _nextId() {
    _nextMsgId++;
    return _nextMsgId.toString();
  }

  Future<Uint8List> _prepareOutboundBody(Uint8List plaintext) async {
    if (plaintext.isEmpty) return Uint8List(0);
    final MessageCrypto? crypto = options.messageCrypto;
    if (crypto == null) return plaintext;
    return crypto.encrypt(plaintext);
  }

  static int _nowMs() => DateTime.now().millisecondsSinceEpoch;

  static Uint8List _toBytes(dynamic data) {
    if (data is Uint8List) return data;
    if (data is List<int>) return Uint8List.fromList(data);
    return Uint8List(0);
  }
}

class _ConnectedPair {
  final WebSocketChannel ws;
  final MicsSession session;

  const _ConnectedPair({required this.ws, required this.session});
}

